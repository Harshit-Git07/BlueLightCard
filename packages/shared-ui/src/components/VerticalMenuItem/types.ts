import { MouseEventHandler } from 'react';
import { IconDefinition } from '@fortawesome/pro-solid-svg-icons';

type VerticalMenuItemProps = {
  label: string;
  selected: boolean;
  onClick?: MouseEventHandler<HTMLElement>;
  href?: string;
  isExternalLink?: boolean;
  icon?: IconDefinition;
};

export default VerticalMenuItemProps;
