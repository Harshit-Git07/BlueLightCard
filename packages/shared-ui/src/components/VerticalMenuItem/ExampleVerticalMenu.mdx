import { Meta, Controls } from '@storybook/blocks';
import * as ExampleVerticalMenu from './ExampleVerticalMenu.stories';

<Meta of={ExampleVerticalMenu} />

# ExampleVerticalMenu

- [Overview](#overview)
- [Props](#props)
- [Implementation](#implementation)
- [Usage Guidelines](#usage-guidelines)
- [Design](#design)

## Overview

This is not a component in the component library,
just a visual guide as to what a list of MenuItems would look like for testing purposes.

## Props

<Controls />

## Implementation
Usage example using tsx

```tsx
type ExampleVerticalMenuProps = {
  numberOfItems: number;
} & VerticalMenuItemProps;
const ExampleVerticalMenu: FC<ExampleVerticalMenuProps> = ({
  numberOfItems,
  ...verticalItemProps
}) => {
  const [selectedItems, setSelectedItems] = useState<Array<number>>([]);

  const itemOnClick = (itemIndex: number) =>
    setSelectedItems((prevState) =>
      prevState.includes(itemIndex)
        ? selectedItems.filter((item) => item !== itemIndex)
        : [...selectedItems, itemIndex]
    );

  return (
    <ul className="w-[263px]">
      {Array(numberOfItems)
        .fill(0)
        .map((_, index) => (
          <VerticalMenuItem
            key={index}
            {...verticalItemProps}
            selected={selectedItems.includes(index)}
            onClick={() => itemOnClick(index)}
          />
        ))}
    </ul>
  );
};
```

## Usage Guidelines

➡️ The parent of the VerticalMenuItem should set the width of the items.

➡️ The parent of the VerticalMenuItem should determine whether the item is selected or not.

## Design

[Figma](https://www.figma.com/design/n2NzB4G3OTR74o3dQ4BOwm/My-Account-%26-Sign-up-Component-library?node-id=66-40592&node-type=canvas&t=EZw5NJ3j8wNNgRKY-0)
