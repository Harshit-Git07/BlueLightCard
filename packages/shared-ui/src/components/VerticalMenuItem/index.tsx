import { FC } from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import NextLink from 'next/link';
import { conditionalStrings } from '../../utils/conditionalStrings';
import VerticalMenuItemProps from './types';

const VerticalMenuItem: FC<VerticalMenuItemProps> = ({
  label,
  selected,
  onClick,
  href,
  isExternalLink,
  icon,
}) => {
  const selectedBorderStyles = 'border-r border-colour-primary dark:border-colour-primary-dark';
  const selectedTextStyles = 'text-colour-primary dark:text-colour-primary-dark';
  const selectedBackgroundStyles =
    'bg-colour-surface-container dark:bg-colour-surface-container-dark';

  const combinedStyles = [
    'h-[50px]',
    'pl-[18px] tablet:pl-[2px]',
    'flex',
    'items-center',
    'text-colour-onSurface',
    'dark:text-colour-onSurface-dark',
    'hover:cursor-pointer',
    'hover:text-colour-primary-hover',
    'hover:dark:text-colour-primary-hover-dark',
    'hover:border-r',
    'hover:border-colour-primary-hover',
    'hover:dark:border-colour-primary-dark',
    conditionalStrings({
      [selectedBorderStyles]: selected,
      [selectedTextStyles]: selected,
      [selectedBackgroundStyles]: selected,
    }),
  ].join(' ');

  const itemIcon = icon ? <FontAwesomeIcon className="ml-2" icon={icon} /> : null;

  return (
    <li className={combinedStyles}>
      {href ? (
        <NextLink
          className="h-full w-full flex items-center hover:opacity-100"
          href={href}
          target={isExternalLink ? '_blank' : undefined}
          aria-label={label}
        >
          {label}
          {itemIcon}
        </NextLink>
      ) : (
        <button className="h-full w-full flex items-center" onClick={onClick} aria-label={label}>
          {label}
          {itemIcon}
        </button>
      )}
    </li>
  );
};

export default VerticalMenuItem;
