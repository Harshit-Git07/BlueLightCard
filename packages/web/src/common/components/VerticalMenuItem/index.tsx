import { FC, MouseEventHandler } from 'react';
import NextLink from 'next/link';

export type Props = {
  label: string;
  selected: boolean; // Parent determines selected state
  onClick?: MouseEventHandler<HTMLElement>;
  href?: string;
};

const VerticalMenuItem: FC<Props> = ({ label, selected, onClick, href }) => {
  const selectedBorderStyles = 'border-r border-colour-primary dark:border-colour-primary-dark';
  const selectedTextStyles = 'text-colour-primary dark:text-colour-primary-dark';
  const selectedBackgroundStyles =
    'bg-colour-surface-container dark:bg-colour-surface-container-dark';

  const className = `h-[50px] pl-[18px] flex items-center hover:cursor-pointer hover:text-colour-primary-hover hover:dark:text-colour-primary-hover-dark hover:border-r hover:border-colour-primary-hover hover:dark:border-colour-primary-dark ${
    selected ? `${selectedBorderStyles} ${selectedTextStyles} ${selectedBackgroundStyles}` : ''
  }`;

  return (
    <li className={className}>
      {href ? (
        <NextLink className="h-full w-full flex items-center" href={href}>
          <p>{label}</p>
        </NextLink>
      ) : (
        <button className="h-full w-full flex items-center" onClick={onClick}>
          <p>{label}</p>
        </button>
      )}
    </li>
  );
};

export default VerticalMenuItem;
