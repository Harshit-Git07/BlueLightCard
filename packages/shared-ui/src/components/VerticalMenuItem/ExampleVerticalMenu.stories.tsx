import { Meta, StoryFn } from '@storybook/react';
import { FC, useState } from 'react';
import VerticalMenuItem, { Props as VerticalMenuItemProps } from './';
import { faArrowUpRightFromSquare, faCoffee, faHome } from '@fortawesome/pro-solid-svg-icons';
import { library } from '@fortawesome/fontawesome-svg-core';
import _L from 'lodash';

type ExampleVerticalMenuProps = {
  numberOfItems: number;
} & VerticalMenuItemProps;
const ExampleVerticalMenu: FC<ExampleVerticalMenuProps> = ({
  numberOfItems,
  ...verticalItemProps
}) => {
  const [selectedItems, setSelectedItems] = useState<number | null>(null);

  const itemOnClick = (itemIndex: number) =>
    setSelectedItems((currentItem) => (currentItem === itemIndex ? null : itemIndex));

  return (
    <ul className="w-[263px]">
      {Array(numberOfItems)
        .fill(0)
        .map((_, index) => (
          <VerticalMenuItem
            key={_L.uniqueId()}
            {...verticalItemProps}
            selected={selectedItems === index}
            onClick={() => itemOnClick(index)}
          />
        ))}
    </ul>
  );
};

const icons = { faArrowUpRightFromSquare, faCoffee, faHome };
library.add(...Object.values(icons));

const iconArgSelect = {
  options: ['none'].concat(...Object.keys(icons)),
  mapping: { none: undefined, ...icons },
  control: {
    type: 'select',
    labels: {
      none: 'No Icon',
      faArrowUpRightFromSquare: 'Link',
      faCoffee: 'Coffee',
      faHome: 'Home',
    },
  },
};

const componentMeta: Meta = {
  title: 'Component System/ExampleVerticalMenu',
  component: ExampleVerticalMenu,
  argTypes: {
    icon: { description: 'Icon name from FontAwesomeSolidIcons (Optional)', ...iconArgSelect },
    href: { description: 'Link to navigate to', type: 'string' },
    numberOfItems: {
      name: 'Number of Items',
      control: {
        type: 'range',
        min: 2,
        step: 1,
      },
    },
  },
};

const DefaultTemplate: StoryFn<ExampleVerticalMenuProps> = (args) => (
  <ExampleVerticalMenu {...args} />
);

export const Default = DefaultTemplate.bind({});
Default.args = {
  numberOfItems: 2,
  label: 'MenuItem',
  onClick: () => 'Item Clicked!',
  href: undefined,
};

export const Hover = DefaultTemplate.bind({});
Hover.args = {
  numberOfItems: 2,
  label: 'MenuItem',
  onClick: undefined,
  href: '/',
};

export const Icon = DefaultTemplate.bind({});
Icon.args = {
  numberOfItems: 2,
  label: 'MenuItem',
  onClick: undefined,
  href: '/',
  icon: faArrowUpRightFromSquare,
};

export default componentMeta;
