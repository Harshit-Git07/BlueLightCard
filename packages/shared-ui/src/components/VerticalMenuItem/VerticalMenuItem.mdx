import { Meta, Canvas, Controls } from '@storybook/blocks';
import * as VerticalMenuItem from './VerticalMenuItem.stories';

<Meta of={VerticalMenuItem} />

# VerticalMenuItem

- [Overview](#overview)
- [Props](#props)
- [Variants and States](#variants-and-states)
- [Implementation](#implementation)
- [Usage Guidelines](#usage-guidelines)
- [Design](#design)

## Overview

A component to be rendered in a menu. When clicked it can either:

- navigate the user to a different page, or
- trigger some functionality when clicked.

## Props

<Controls />

## Variants and States

### Default

<Canvas of={VerticalMenuItem.Default} />

### Hover

<Canvas of={VerticalMenuItem.Hover} />

### Selected

<Canvas of={VerticalMenuItem.Selected} />

## Implementation

Usage example using tsx

```tsx
type ExampleVerticalMenuProps = {
  numberOfItems: number;
} & VerticalMenuItemProps;
const ExampleVerticalMenu: FC<ExampleVerticalMenuProps> = ({
  numberOfItems,
  ...verticalItemProps
}) => {
  const [selectedItems, setSelectedItems] = useState<Array<number>>([]);

  const itemOnClick = (itemIndex: number) =>
    setSelectedItems((prevState) =>
      prevState.includes(itemIndex)
        ? selectedItems.filter((item) => item !== itemIndex)
        : [...selectedItems, itemIndex]
    );

  return (
    <ul className="w-[263px]">
      {Array(numberOfItems)
        .fill(0)
        .map((_, index) => (
          <VerticalMenuItem
            key={index}
            {...verticalItemProps}
            selected={selectedItems.includes(index)}
            onClick={() => itemOnClick(index)}
          />
        ))}
    </ul>
  );
};
```

## Usage Guidelines

➡️ The parent of the VerticalMenuItem should set the width of the items.

➡️ The parent of the VerticalMenuItem should determine whether the item is selected or not.

➡️ The `href` prop has priority over the `onClick` prop if both are supplied.

## Design

[Figma](https://www.figma.com/design/n2NzB4G3OTR74o3dQ4BOwm/My-Account-%26-Sign-up-Component-library?node-id=66-40592&node-type=canvas&t=EZw5NJ3j8wNNgRKY-0)
