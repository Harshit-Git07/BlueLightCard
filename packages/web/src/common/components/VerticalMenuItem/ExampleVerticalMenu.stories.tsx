import { Meta, StoryFn } from '@storybook/react';
import { FC, useState } from 'react';
import VerticalMenuItem, { Props as VerticalMenuItemProps } from './';
import _L from 'lodash';

type ExampleVerticalMenuProps = {
  numberOfItems: number;
} & VerticalMenuItemProps;
const ExampleVerticalMenu: FC<ExampleVerticalMenuProps> = ({
  numberOfItems,
  ...verticalItemProps
}) => {
  const [selectedItems, setSelectedItems] = useState<number | null>(null);

  const itemOnClick = (itemIndex: number) =>
    setSelectedItems((currentItem) => (currentItem === itemIndex ? null : itemIndex));

  return (
    <ul className="w-[263px]">
      {Array(numberOfItems)
        .fill(0)
        .map((_, index) => (
          <VerticalMenuItem
            key={_L.uniqueId()}
            {...verticalItemProps}
            selected={selectedItems === index}
            onClick={() => itemOnClick(index)}
          />
        ))}
    </ul>
  );
};

const componentMeta: Meta = {
  title: 'Component System/ExampleVerticalMenu',
  component: ExampleVerticalMenu,
  argTypes: {
    numberOfItems: {
      name: 'Number of Items',
      control: {
        type: 'range',
        min: 2,
        step: 1,
      },
    },
  },
};

const DefaultTemplate: StoryFn<ExampleVerticalMenuProps> = (args) => (
  <ExampleVerticalMenu {...args} />
);

export const OnClickVariant = DefaultTemplate.bind({});
OnClickVariant.args = {
  numberOfItems: 2,
  label: 'MenuItem',
  onClick: () => 'Item Clicked!',
  href: undefined,
};

export const HrefVariant = DefaultTemplate.bind({});
HrefVariant.args = {
  numberOfItems: 2,
  label: 'MenuItem',
  onClick: undefined,
  href: '/',
};

export default componentMeta;
