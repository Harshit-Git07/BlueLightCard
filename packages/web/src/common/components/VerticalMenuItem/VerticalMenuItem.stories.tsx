import { Meta, StoryFn } from '@storybook/react';
import VerticalMenuItem from './';

const componentMeta: Meta<typeof VerticalMenuItem> = {
  title: 'Component System/VerticalMenuItem',
  component: VerticalMenuItem,
};

const DefaultTemplate: StoryFn<typeof VerticalMenuItem> = (args) => (
  <ul>
    <VerticalMenuItem {...args} />
  </ul>
);

export const OnClick = DefaultTemplate.bind({});

OnClick.args = {
  label: 'Default Menu Item',
  onClick: () => "I've been Clicked!",
};

export const HRef = DefaultTemplate.bind({});

HRef.args = {
  label: 'Default Menu Item',
  href: '/',
};

export const Selected = DefaultTemplate.bind({});

Selected.args = {
  label: 'Selected Menu Item',
  onClick: () => "I've been Clicked!",
  selected: true,
};

export default componentMeta;
