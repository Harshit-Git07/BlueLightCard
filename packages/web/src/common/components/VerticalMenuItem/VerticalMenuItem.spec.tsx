import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';
import VerticalMenuItem, { Props } from './';
import userEvent from '@testing-library/user-event';

describe('VerticalMenuItem component', () => {
  const defaultOnClickProps: Props = {
    label: 'test',
    onClick: () => undefined,
    selected: false,
  };

  const defaultHRefProps: Props = {
    label: 'test',
    href: '/',
    selected: false,
  };

  // smoke test
  it('should render without error', async () => {
    const { baseElement: onClickElement } = render(<VerticalMenuItem {...defaultOnClickProps} />);
    expect(onClickElement).toBeTruthy();
    const { baseElement: hrefElement } = render(<VerticalMenuItem {...defaultHRefProps} />);
    expect(hrefElement).toBeTruthy();
  });

  it('triggers onClick fn when clicked', async () => {
    const mockFn = jest.fn();
    render(<VerticalMenuItem {...defaultOnClickProps} onClick={mockFn} />);

    const textToClick = screen.getByText('test');
    await userEvent.click(textToClick);

    expect(mockFn).toHaveBeenCalled();
  });

  it('renders a link when passed href', async () => {
    render(<VerticalMenuItem {...defaultHRefProps} />);

    const linkItem = screen.getByRole('link');

    expect(linkItem).toBeTruthy();
  });

  it('matches snapshot in its default state', async () => {
    const { container } = render(<VerticalMenuItem {...defaultOnClickProps} />);

    expect(container).toMatchSnapshot();
  });

  it('matches snapshot in its selected state', async () => {
    const { container } = render(<VerticalMenuItem {...defaultOnClickProps} selected />);

    expect(container).toMatchSnapshot();
  });
});
