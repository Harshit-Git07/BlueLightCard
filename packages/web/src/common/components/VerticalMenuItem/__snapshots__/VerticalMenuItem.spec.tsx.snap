// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VerticalMenuItem component matches snapshot in its default state 1`] = `
<div>
  <li
    class="h-[50px] pl-[18px] flex items-center hover:cursor-pointer hover:text-colour-primary-hover hover:dark:text-colour-primary-hover-dark hover:border-r hover:border-colour-primary-hover hover:dark:border-colour-primary-dark "
  >
    <button
      class="h-full w-full flex items-center"
    >
      <p>
        test
      </p>
    </button>
  </li>
</div>
`;

exports[`VerticalMenuItem component matches snapshot in its selected state 1`] = `
<div>
  <li
    class="h-[50px] pl-[18px] flex items-center hover:cursor-pointer hover:text-colour-primary-hover hover:dark:text-colour-primary-hover-dark hover:border-r hover:border-colour-primary-hover hover:dark:border-colour-primary-dark border-r border-colour-primary dark:border-colour-primary-dark text-colour-primary dark:text-colour-primary-dark bg-colour-surface-container dark:bg-colour-surface-container-dark"
  >
    <button
      class="h-full w-full flex items-center"
    >
      <p>
        test
      </p>
    </button>
  </li>
</div>
`;
